### PR TITLE
storage: add interfaces for copysets

### PR DESCRIPTION
**Pull request by mvijaykarthik**
Copysets significantly reduce the probability of data loss
in the presence of multi node failures. The RFC for copysets
is presented in https://github.com/cockroachdb/cockroach/pull/32816.

This commit adds interfaces and protos for copysets.
The implementation added later will have two independent tracks:
1. Assignment and persistance of store_id-copyset_id mapping.
2. Rebalancer changes to rebalance replicas into copysets.

Release note: None

### CODE REVIEW

**Review by tbg — CHANGES_REQUESTED** [review]
> Reviewed 1 of 7 files at r1, 6 of 6 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32954)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mvijaykarthik)

___
*[pkg/storage/copysets.go, line 15 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/32954#-LXdVdmyDDZsFAuTp4Ch:-LXdVdmyDDZsFAuTp4Ci:b-sxzmvt) ([raw file](https://github.com/cockroachdb/cockroach/blob/945ebd9e4a3123c998c3eae3dc7f1aec2eb301d8/pkg/storage/copysets.go#L15)):*
> ```Go
> // permissions and limitations under the License.
> 
> package storage
> ```

Please move both this and the protos into a `storage/copysets` package. You'll have to do something in the `Makefile` to make this all work, I think. It should be similar to other packages like `ctpb`.

Please add a `doc.go` with a high-level description of copysets (i.e. what is it, how do they work, what do we implement) and a link to the paper (this should already be in the RFC, definitely don't rewrite this from scratch).

File: pkg/roachpb/metadata.go [hunk 1]
```diff
@@ -52,6 +52,11 @@ func (n StoreID) String() string {
 	return strconv.FormatInt(int64(n), 10)
 }
 
+// CopysetID is a custom type for a cockroach copyset ID.
+// Copyset is a group of stores where a range should be contained within
+// if copyset based rebalancing is enabled.
+type CopysetID int32
```
Comment (tbg) [inline pkg/roachpb/metadata.go]: Does this make sense in `roachpb`? Right now it seems that it would live in `copysets` as well.

**Review by tbg — APPROVED** [review]
Thanks for the rearrangements. LGTM mod the comment.

**Review by mvijaykarthik — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32954)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mvijaykarthik and @tbg)

___
*[pkg/roachpb/metadata.go, line 58 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/32954#-LTBdRmdR2SNYo6L7qlU-r3-58:-LY2C5Rq2Qu1n_GryhMJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0706898d8b6d5c897256826b41ec84537f1e962/pkg/roachpb/metadata.go#L58)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Does this make sense in `roachpb`? Right now it seems that it would live in `copysets` as well.
</blockquote></details>

Done.

**Comment by tbg** [comment]
Thank you!

bors r+

**Pull request merged**